### PR TITLE
Create and edit routes with multiple services

### PR DESCRIPTION
**Pull request by spadgett**
https://trello.com/c/v3jASHVU

screen shot 2016-07-08 at 2 38 40 pm

/cc @jwforres

### CODE REVIEW

**Comment by spadgett** [comment]
@jwforres thoughts on the pie chart?

screen shot 2016-07-08 at 4 02 31 pm

**Comment by spadgett** [comment]
@jwforres updated screenshot

screen shot 2016-07-11 at 10 39 28 am

**Comment by spadgett** [comment]
I should remove "Service" from each legend item

**Comment by spadgett** [comment]
@jwforres PTAL

**Comment by jwforres** [comment]
I like this much better than what you had this morning 👍

File: app/scripts/controllers/edit/route.js [hunk 1]
```diff
@@ -73,16 +88,13 @@ angular.module('openshiftConsole')
...
-          var serviceName = _.get($scope, 'routing.service.metadata.name');
+          var serviceName = _.get($scope, 'routing.to.service.metadata.name');
           _.set(route, 'spec.to.name', serviceName);
-
-          // Remove the host.generated annotation if the host was edited.
-          if (_.get(route, ['metadata', 'annotations', 'openshift.io/host.generated']) === 'true' &&
-              _.get(route, 'spec.host') !== $scope.routing.host) {
-            delete route.metadata.annotations['openshift.io/host.generated'];
+          var weight = _.get($scope, 'routing.to.weight');
+          if (!isNaN(weight)) {
```
Comment (jwforres) [inline app/scripts/controllers/edit/route.js]: are you not doing number validation on the inputs, or is this just an extra sanity check

File: app/scripts/directives/oscRouting.js [hunk 2]
```diff
@@ -79,13 +80,20 @@ angular.module("openshiftConsole")
...
+          updatePortOptions(newValue);
           // Don't overwrite the target port when editing an existing route unless the user picked a
           // different service.
           if (newValue !== oldValue || !scope.route.targetPort) {
             scope.route.targetPort = _.get(scope, 'route.portOptions[0].port');
           }
+
+          if (scope.services) {
+            // Update the options for alternate services.
+            scope.alternateServiceOptions = _.reject(scope.services, function(service) {
```
Comment (jwforres) [inline app/scripts/directives/oscRouting.js]: what happens if they have already added the service as an alternate and they switch it to be the primary?

File: app/scripts/directives/routeServicePie.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      template: '<div ng-attr-id="{{chartId}}"></div>',
+      link: function($scope) {
+        var chart, config;
+
+        $scope.chartId = _.uniqueId('route-service-chart-');
+
+        config = {
+          bindto: '#' + $scope.chartId,
+          color: {
+            pattern: [$.pfPaletteColors.blue, $.pfPaletteColors.orange, $.pfPaletteColors.green, $.pfPaletteColors.red]
```
Comment (jwforres) [inline app/scripts/directives/routeServicePie.js]: what does c3 do if it needs more colors, start adjusting the shade of the color?

File: app/views/directives/osc-routing-service.html [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+             type="number"
+             required
+             min="1"
+             max="256"
+             ng-pattern="/^\d+$/"
+             class="form-control"
+             aria-describedby="{{id}}-weight-help">
+      <div>
+        <span id="{{id}}-weight-help" class="help-block">
+          Weight as an number between 1 and 256 that specifies the relative weight against
```
Comment (jwforres) [inline app/views/directives/osc-routing-service.html]: 'Weight is a number...'

File: app/views/directives/osc-routing-service.html [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          other route services.
+        </span>
+        <div ng-if="routingServiceForm.weight.$dirty && routingServiceForm.weight.$invalid" class="has-error">
+          <div ng-if="routingServiceForm.weight.$error.number" class="help-block">
+            Must be a number.
+          </div>
+          <div ng-if="routingServiceForm.weight.$error.min" class="help-block">
+            Must greater than or equal to 1.
```
Comment (jwforres) [inline app/views/directives/osc-routing-service.html]: 'Must be...'

File: app/views/directives/osc-routing-service.html [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      </select>
+      <div>
+        <span ng-attr-id="{{id}}-service-help" class="help-block">
+          <span ng-if="!isAlternate">Service to route to.</span>
+          <span ng-if="isAlternate">Alternate service for route traffic.</span>
+        </span>
+      </div>
+      <div ng-if="(services | hashSize) === 0" class="has-error">
+        <span class="help-block">
+          There are no <span ng-if="is-alternate">additional</span> services in your project to expose
```
Comment (jwforres) [inline app/views/directives/osc-routing-service.html]: what if there is only one service in the project?  you can route to that single service in the to, but would have no more to pick from to be "additional",  services|hashSize would be zero and this help block wouldnt show, unless im missing something

File: app/scripts/controllers/edit/route.js (hunk 1 above)
Comment (spadgett) [inline app/scripts/controllers/edit/route.js]: It's not set for routes with only one service.

File: app/scripts/directives/oscRouting.js (hunk 2 above)
Comment (spadgett) [inline app/scripts/directives/oscRouting.js]: The selected alternate option is removed, so dropdown has nothing selected. It's a required field, so they have to pick another before saving.

File: app/scripts/directives/routeServicePie.js (hunk 3 above)
Comment (spadgett) [inline app/scripts/directives/routeServicePie.js]: The colors start repeating (5 would be blue, 6 orange, etc.)

File: app/views/directives/osc-routing-service.html (hunk 6 above)
Comment (spadgett) [inline app/views/directives/osc-routing-service.html]: We don't show the option to split across services at all if there is only one, but I think the logic is right? scope.services here is the list of additional services _excluding_ the primary if this is an alternate.

Comment (jwforres) [inline app/views/directives/osc-routing-service.html]: ah didnt realize the scope of what this form was, got it

**Comment by jwforres** [comment]
[merge]

**Pull request merged**